intf/intf.go

```diff
@@ -154,3 +168,56 @@ func AddIP(intf string, addr *net.IPNet) error {
 	}
 	return nil
 }
+
+// AwaitARP waits for the IPv4 address addr to be resolved via ARP. It uses the supplied
+// context to determine whether it should continue awaiting the entry. If the ARP entry
+// is resolved, it is returned directly.
+func AwaitARP(ctx context.Context, addr net.IP) (net.HardwareAddr, error) {
+	neighs, err := accessor.ARPList()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve list of ARP entries, %v", err)
+	}
+
+	for _, n := range neighs {
+		if addr.Equal(n.IP) {
+			return n.MAC, nil
+		}
+	}
+
+	updates := make(chan ARPUpdate, 1)
+	result := make(chan ARPUpdate, 1)
+	done := make(chan struct{})
+
+	go func(updates chan ARPUpdate, done chan struct{}) {
+		for {
+			select {
+			case upd := <-updates:
+				if upd.Type == ARPAdd {
+					if addr.Equal(upd.Neigh.IP) {
+						result <- upd
+					}
+				}
+			case <-done:
```

Comment:
so the usage pattern in the ARPSubscribe will be that it'll ARP out for the given IP, and upon receiving responses, will write them to the updates channel? then at some point write to done to tell AwaitARP that no more responses are coming? 

Why do we not return if `addr.Equal(upd.Neigh.IP)`, meaning we successfully found the MAC address of the requested IP?

Reply:
It doesn't trigger ARP, but it will wait until ARP for that address is resolved. It writes the updates that it receives for that IP to the channel so that the caller can do something with that now-resolved ARP entry.

It doesn't write to `done`, but rather uses that to determine whether it should keep listening - so we get this goroutine to return when we write to `done` on line 212.

I had originally had this as a more generic listener, where it would allow you to continue to listen, you're right that this can be simplified to just return the goroutine when it gets the first update, given the caller does that.

We still write to `done` on 212 because that tells the accessor's underlying `ARPSubscribe` to exit.